File: Leetcode/linkedlist/138_CopyListWithRandomPointer/Solution.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val, next, random):
        self.val = val
        self.next = next
        self.random = random
"""
class Solution(object):
    def copyRandomList(self, head):
        """
        :type head: Node
        :rtype: Node
        """
        if not head:
            return None
        p = head
        while p:
            node = Node(p.val, None, None)
            node.next = p.next
            p.next = node
            p = p.next.next
        p = head
        while p:
            if p.random:
                p.next.random = p.random.next
            p = p.next.next
        newhead = head.next
        pold = head
        pnew = newhead
        while pnew.next:
            pold.next = pnew.next
            pold = pold.next
            pnew.next = pold.next
            pnew = pnew.next
        pold.next = None
        pnew.next = None
        return newhead

File: Leetcode/linkedlist/138_CopyListWithRandomPointer/Solution.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node next;
    public Node random;

    public Node() {}

    public Node(int _val,Node _next,Node _random) {
        val = _val;
        next = _next;
        random = _random;
    }
};
*/
class Solution {
    public Node copyRandomList(Node phead) {
        if(phead==null) return null;
        Node res = new Node(phead.val);
        Map<Node, Node> map = new HashMap<>();
        Node temp = phead;
        temp = phead;
        Node temp2 = res;
        while(temp!=null){
            map.put(temp, temp2);
            temp = temp.next;
            temp2.next = (temp==null)? null: new Node(temp.val);
            temp2 = (temp2.next==null)? temp2: temp2.next;
        }
        temp = phead;
        temp2 = res;
        while(temp!=null){
            Node random = map.get(temp.random);
            temp2.random = random;
            temp = temp.next;
            temp2 = temp2.next;
        }
        return res;
    }
}
